### PR TITLE
Multiple parameters support

### PR DESCRIPTION
**Pull request by akosyakov**
Includes https://github.com/eclipse/lsp4j/pull/89

### CODE REVIEW

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/JsonRpcMethod.java [hunk 1]
```diff
@@ -62,26 +60,24 @@ public String toString() {
...
 	}
 
-	@Override
 	public int hashCode() {
 		final int prime = 31;
 		int result = 1;
 		result = prime * result + (isNotification ? 1231 : 1237);
 		result = prime * result + ((methodName == null) ? 0 : methodName.hashCode());
-		result = prime * result + ((parameterType == null) ? 0 : parameterType.hashCode());
+		result = prime * result + ((parameterTypes == null) ? 0 : parameterTypes.hashCode());
```
Comment (spoenemann) [inline org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/JsonRpcMethod.java]: Would it make sense to use `Arrays.HashCode(..)` here?

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/JsonRpcMethod.java [hunk 2]
```diff
@@ -97,10 +93,10 @@ public boolean equals(Object obj) {
 				return false;
 		} else if (!methodName.equals(other.methodName))
 			return false;
-		if (parameterType == null) {
-			if (other.parameterType != null)
+		if (parameterTypes == null) {
+			if (other.parameterTypes != null)
 				return false;
-		} else if (!parameterType.equals(other.parameterType))
+		} else if (!parameterTypes.equals(other.parameterTypes))
```
Comment (spoenemann) [inline org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/JsonRpcMethod.java]: Would it make sense to use `Arrays.equals(..)` here?

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapterFactory.java [hunk 3]
```diff
@@ -122,24 +125,62 @@ public Message read(JsonReader in) throws IOException {
...
+		
+		protected Object fromJson(JsonElement element, Type type) {
+			if (isNull(element)) {
+				return null;
+			}
+			if (isNullOrVoidType(type)) {
+				return element;
+			}
+			Object value = gson.fromJson(element, type);
+			if (isNull(value)) {
```
Comment (spoenemann) [inline org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapterFactory.java]: The result of `gson.fromJson(element, type)` will never be a JsonNull?

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapterFactory.java [hunk 4]
```diff
@@ -122,24 +125,62 @@ public Message read(JsonReader in) throws IOException {
...
+			if (isNull(rawParams)) {
+				return null;
 			}
-			return null;
+			Type[] parameterTypes = getParameterTypes(method);
+			if (parameterTypes.length == 1) {
+				return fromJson(rawParams, parameterTypes[0]);
+			}
+			if (parameterTypes.length > 1 && rawParams instanceof JsonArray) {
+				List<Object> parameters = new ArrayList<Object>();
```
Comment (spoenemann) [inline org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapterFactory.java]: You could initialize the ArrayList with the size of the JsonArray.

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapterFactory.java [hunk 5]
```diff
@@ -88,12 +96,7 @@ public Message read(JsonReader in) throws IOException {
 					break;
 				}
 				case "params": {
-					Type parameterType = getParameterType(method);
-					if (isNullOrVoidType(parameterType)) {
-						params = new JsonParser().parse(in);
-					} else {
-						params = gson.fromJson(in, parameterType);
-					}
+					rawParams = new JsonParser().parse(in);
```
Comment (spoenemann) [inline org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapterFactory.java]: With this change we always create a JsonElement tree, which is less efficient than creating the parameters directly from the stream. Can we still support both without too much code overhead?

**Comment by akosyakov** [comment]
@spoenemann i've addressed your comments, please review again.

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapterFactory.java [hunk 6]
```diff
@@ -128,11 +128,47 @@ public Message read(JsonReader in) throws IOException {
...
+					Type parameterType = index < parameterTypes.length ? parameterTypes[index] : null;
+					Object parameter = fromJson(in, parameterType);
+					parameters.add(parameter);
+					index++;
+				}
+				while (index < parameterTypes.length) {
+					parameters.add(null);
+					index++;
+				}
+				if (in.peek() == JsonToken.END_ARRAY) {
```
Comment (spoenemann) [inline org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapterFactory.java]: I don't think this check is necessary. Same for `in.peek() != JsonToken.END_ARRAY` in l. 146. The JsonReader should guarantee that the next token is `END_ARRAY` if and only if `in.hasNext()` is false.

**Comment by spoenemann** [comment]
Found a minor issue, otherwise good work! 👍

**Pull request merged**